commands/http/handler.go

```diff
@@ -39,16 +48,31 @@ var mimeTypes = map[string]string{
 	cmds.Text: "text/plain",
 }
 
-func NewHandler(ctx cmds.Context, root *cmds.Command, origin string) *Handler {
+func NewHandler(ctx cmds.Context, root *cmds.Command, allowedOrigin string) *Handler {
 	// allow whitelisted origins (so we can make API requests from the browser)
-	if len(origin) > 0 {
-		log.Info("Allowing API requests from origin: " + origin)
+	if len(allowedOrigin) > 0 {
+		log.Info("Allowing API requests from origin: " + allowedOrigin)
 	}
 
-	return &Handler{ctx, root, origin}
+	// Create a handler for the API.
+	internal := internalHandler{ctx, root}
+
+	// Create a CORS object for wrapping the internal handler.
+	c := cors.New(cors.Options{
+		AllowedMethods: []string{"GET", "POST", "PUT"},
+
+		// use AllowOriginFunc instead of AllowedOrigins because we want to be
+		// restrictive by default.
+		AllowOriginFunc: func(origin string) bool {
+			return (allowedOrigin == "*") || (origin == allowedOrigin)
+		},
```

Comment:
is this fixing that one case where the CORS reply has to be the origin of the request, instead of `*` ?